examples/llm/src/models/param_init_fns.py

```diff
@@ -0,0 +1,161 @@
+            torch.nn.init.zeros_(module.in_proj_bias)
+        if module.bias_k is not None:
+            torch.nn.init.zeros_(module.bias_k)
+        if module.bias_v is not None:
+            torch.nn.init.zeros_(module.bias_v)
+
+        # out proj
+        init_fn(module.out_proj.weight)
+        if div_is_residual and getattr(module.out_proj, '_is_residual', False):
+            with torch.no_grad():
+                module.out_proj.weight.div_(math.sqrt(2 * cfg.n_layers))
+        if module.out_proj.bias is not None:
+            torch.nn.init.zeros_(module.out_proj.bias)
+
+
+def baseline_param_init_fn(module, cfg):
+    init_fn = partial(torch.nn.init.normal_, mean=0.0, std=cfg.init_std)
+    generic_param_init_fn(module, cfg, init_fn)
+
+
+def kaiming_uniform_param_init_fn(module, cfg):
+    # set nn.init.kaiming_uniform_ defaults
+    # note: not many ppl modify the defaults but maybe we should...
+    init_gain = cfg.get('init_gain', 0)
+    fan_mode = cfg.get('fan_mode', 'fan_in')
+    init_nonlinearity = cfg.get('init_nonlinearity', 'leaky_relu')
+
+    kaiming_uniform_ = partial(nn.init.kaiming_uniform_,
+                               a=init_gain,
+                               mode=fan_mode,
+                               nonlinearity=init_nonlinearity)
+
+    generic_param_init_fn(module, cfg, kaiming_uniform_)
+
+
+def kaiming_normal_param_init_fn(module, cfg):
+    # set torch.nn.init.kaiming_normal_ defaults
+    # note: not many ppl modify the defaults but maybe we should...
+    init_gain = cfg.get('init_gain', 0)
+    fan_mode = cfg.get('fan_mode', 'fan_in')
+    init_nonlinearity = cfg.get('init_nonlinearity', 'leaky_relu')
+
+    kaiming_normal_ = partial(torch.nn.init.kaiming_normal_,
+                              a=init_gain,
+                              mode=fan_mode,
+                              nonlinearity=init_nonlinearity)
+
+    generic_param_init_fn(module, cfg, kaiming_normal_)
+
+
+def xavier_uniform_param_init_fn(module, cfg):
+    # set nn.init.xavier_uniform_ defaults
+    # note: not many ppl modify the defaults but maybe we should...
+    init_gain = cfg.get('init_gain', 1.0)
+    xavier_uniform_ = partial(torch.nn.init.xavier_uniform_, gain=init_gain)
+
+    generic_param_init_fn(module, cfg, xavier_uniform_)
+
+
+def xavier_normal_param_init_fn(module, cfg):
+    # set nn.init.xavier_normal_ defaults
+    # note: not many ppl modify the defaults but maybe we should...
+    init_gain = cfg.get('init_gain', 1.0)
+    xavier_normal_ = partial(torch.nn.init.xavier_normal_, gain=init_gain)
+
+    generic_param_init_fn(module, cfg, xavier_normal_)
+
+
+MODEL_INIT_REGISTRY = {
+    'default': torch_defualt_param_init_fn,
+    'baseline': baseline_param_init_fn,
+    'kaiming_uniform': kaiming_uniform_param_init_fn,
+    'kaiming_normal': kaiming_normal_param_init_fn,
+    'xavier_uniform': xavier_uniform_param_init_fn,
+    'xavier_normal_': xavier_normal_param_init_fn,
```

Comment:
extra underscore here? or why is this one special?

Reply:
Underscore to imply in-place since the weights are modified in-place; new weights are not generated; should have been done to all the fns.
Should I make all of the fn with underscores (cuz in-place) or just not use that convention.

see PyTorch example for in-place using underscore: https://pytorch.org/docs/stable/nn.init.html#torch.nn.init.uniform_ ie all the torch init fns

Reply:
ah, got it. yeah lets go with all of them having underscores.